pkg/updatestrategy/drain.go

```diff
@@ -0,0 +1,450 @@
+		}
+	}
+	return false
+}
+
+func findSiblingPods(logger *log.Entry, pod v1.Pod, allPods []v1.Pod, allPdbs []policy.PodDisruptionBudget) []v1.Pod {
+	var siblingSelectors []labels.Selector
+	for _, pdb := range allPdbs {
+		selector, err := metav1.LabelSelectorAsSelector(pdb.Spec.Selector)
+		if err != nil {
+			logger.Debugf("pdb %s/%s has an invalid selector: %v", pod.GetNamespace(), pdb.GetName(), err)
+			continue
+		}
+		if selector.Matches(labels.Set(pod.Labels)) {
+			siblingSelectors = append(siblingSelectors, selector)
+		}
+	}
+
+	var result []v1.Pod
+	for _, candidate := range allPods {
+		if candidate.GetName() == pod.GetName() {
+			continue
+		}
+
+		for _, selector := range siblingSelectors {
+			if selector.Matches(labels.Set(candidate.Labels)) {
+				result = append(result, candidate)
+				break
+			}
+		}
+	}
+
+	return result
+}
+
+func isPDBViolation(err error) bool {
+	return apiErrors.IsTooManyRequests(err) || strings.Contains(err.Error(), multiplePDBsErrMsg)
+}
+
+var deletePod = func(client kubernetes.Interface, logger *log.Entry, pod v1.Pod) error {
+	err := client.CoreV1().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{
+		GracePeriodSeconds: pod.Spec.TerminationGracePeriodSeconds,
+	})
+	if err != nil {
+		logger.Errorf("Failed to delete pod: %v", err)
+		return err
+	}
+
+	// wait for pod to be terminated and gone from the node.
+	err = waitForPodTermination(client, pod)
+	if err != nil {
+		logger.Warnf("Pod not terminated within grace period: %s", err)
+	}
+
+	logger.Info("Pod deleted")
+	return nil
+
+}
+
+// evictPod tries to evict a pod from a node.
+// Note: this is defined as a variable so it can be easily mocked in tests.
+var evictPod = func(client kubernetes.Interface, logger *log.Entry, pod v1.Pod) error {
+	localLogger := logger.WithFields(log.Fields{
+		"ns":   pod.Namespace,
+		"pod":  pod.Name,
+		"node": pod.Spec.NodeName,
+	})
+
+	updated, err := client.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if updated.Status.Phase == v1.PodSucceeded || updated.Status.Phase == v1.PodFailed {
+		// Completed, just ignore
+		return nil
+	}
+
+	if updated.Spec.NodeName != pod.Spec.NodeName {
+		// Already evicted
+		return nil
+	}
+
+	eviction := &policy.Eviction{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+		},
+		DeleteOptions: &metav1.DeleteOptions{
+			GracePeriodSeconds: pod.Spec.TerminationGracePeriodSeconds,
+		},
+	}
+
+	err = client.CoreV1().Pods(pod.Namespace).Evict(eviction)
+	if err != nil {
+		return err
+	}
+	localLogger.Info("Evicting pod")
+
+	// wait for the pod to be actually evicted and gone from the node.
+	// It has TerminationGracePeriodSeconds time to clean up.
+	start := time.Now().UTC()
+	err = waitForPodTermination(client, pod)
+	if err != nil {
+		localLogger.Warnf("Pod not terminated within grace period: %s", err)
```

Comment:
info or error, not warn, which should be only used for deprecations

Reply:
See above, in my opinion warnings are fine.